File: network/java/src/main/java/org/aiddl/network/AiddlGrpcServer.java
package org.aiddl.network;

import java.io.IOException;
import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;

import org.aiddl.core.function.FunctionRegistry;
import org.aiddl.core.interfaces.Function;
import org.aiddl.core.parser.Parser;
import org.aiddl.core.representation.Term;
import org.aiddl.network.AiddlOuterClass.FunctionCallRequest;
import org.aiddl.network.AiddlOuterClass.FunctionCallResult;
import org.aiddl.network.AiddlOuterClass.FunctionCallResult.Builder;

import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.stub.StreamObserver;

public class AiddlGrpcServer {
	private final Server server;

	public AiddlGrpcServer(int port, FunctionRegistry freg, boolean allow_remote_registry) throws IOException {
		if ( allow_remote_registry ) {
			freg.addFunction(Term.sym("org.aiddl.network.register-remote-function"), new GrpcFunctionLoader(freg));
		}
		
		server = ServerBuilder.forPort(port).addService(new AiddlService(freg))
				.build();
	}

	/** Start serving requests. */
	public void start() throws IOException {
		server.start();
//		logger.info("Server started, listening on " + port);
		Runtime.getRuntime().addShutdownHook(new Thread() {
			@Override
			public void run() {
				// Use stderr here since the logger may have been reset by its JVM shutdown hook.
				System.err.println("*** shutting down gRPC server since JVM is shutting down");
				try {
					AiddlGrpcServer.this.stop();
				} catch (InterruptedException e) {
					e.printStackTrace(System.err);
				}
				System.err.println("*** server shut down");
			}
		});
	}

	/** Stop serving requests and shutdown resources. */
	public void stop() throws InterruptedException {
		if (server != null) {
			server.shutdown().awaitTermination(30, TimeUnit.SECONDS);
		}
	}

	private static class AiddlService extends AiddlGrpc.AiddlImplBase {
		private FunctionRegistry freg;

		public AiddlService( FunctionRegistry  freg ) {
			this.freg = freg;
		}

		/**
		 *
		 * @param request the requested location for the feature.
		 * @param responseObserver the observer that will receive the feature at the requested point.
		 */
		@Override
		public void functionCall(FunctionCallRequest request, StreamObserver<FunctionCallResult> responseObserver) {
			responseObserver.onNext(runPlanner(request));
			responseObserver.onCompleted();
		}

		private FunctionCallResult runPlanner(FunctionCallRequest problem) {
			Term function_uri = Parser.ParseTerm(problem.getFunctionUri());
			Term function_arg = Parser.ParseTerm(problem.getArg());

			System.out.println("======= Request =======");
			System.out.println(function_uri);
			System.out.println(function_arg);
			
			Function f = this.freg.getFunction(function_uri);
			Term r = f.apply(function_arg);
			Builder s; 
			System.out.println(r);
			
			s = FunctionCallResult.newBuilder().setStatus(0);

			s.setResult(r.toString());

			return s.build();
		}
	}	
}

File: util/python/src/aiddl_util/grpc/aiddl_grpc_server.py
from concurrent import futures

import grpc

from aiddl_core.representation.sym import Sym
from aiddl_core.parser import parser

from aiddl_util.grpc.grpc_function import GrpcFunctionLoader
import aiddl_util.grpc.generated.aiddl_pb2 as aiddl_pb2
import aiddl_util.grpc.generated.aiddl_pb2_grpc as aiddl_pb2_grpc

LOADER_URI = Sym("org.aiddl.network.register-remote-function")


class AiddlServicer(aiddl_pb2_grpc.AiddlServicer):
    def __init__(self, port, freg, allow_remote_registry=False):
        self.port = port
        self.freg = freg
        if allow_remote_registry:
            self.freg.add_function(LOADER_URI,
                                   GrpcFunctionLoader(self.freg))

    def FunctionCall(self, request, context):
        uri = parser.parse_term(request.function_uri)
        args = parser.parse_term(request.arg)
        # print("Request to call:", uri, "with:", args)
        f = self.freg.get_function(uri)
        r = f(args)
        # print("Result:", r)
        result = aiddl_pb2.FunctionCallResult(status=0, result=str(r))
        return result

    def start(self):
        self.server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
        aiddl_pb2_grpc.add_AiddlServicer_to_server(
            self, self.server)
        self.server.add_insecure_port('0.0.0.0:%d' % (self.port))
        self.server.start()

    def wait_for_termination(self):
        self.server.wait_for_termination()
